ui/src/app/components/popups.tsx

```diff
@@ -0,0 +1,285 @@
+    const {children, side, elementRef, dimensions: {target, element, viewport}} = this.props;
+    const {side: finalSide, position} =
+      computePopupPosition({ side, target, element, viewport, gap: 10 });
+    const getNotchPosition = () => {
+      const left = clamp(12, element.width - 12,
+        (target.left + target.right) / 2 - position.left
+      );
+      const top = clamp(12, element.height - 12,
+        (target.top + target.bottom) / 2 - position.top
+      );
+      return switchCase(finalSide,
+        ['top', () => ({ bottom: 0, left, transform: 'rotate(180deg)' })],
+        ['bottom', () => ({ top: 0, left })],
+        ['left', () => ({ right: 0, top, transform: 'rotate(90deg)' })],
+        ['right', () => ({ left: 0, top, transform: 'rotate(270deg)' })]
+      );
+    };
+    return <PopupPortal>
+      <div
+        ref={elementRef}
+        style={{
+          transform: `translate(${position.left}px, ${position.top}px)`,
+          ...styles.tooltip
+        } as React.CSSProperties}
+      >
+        {children}
+        <svg viewBox='0 0 2 1' style={{...getNotchPosition(), ...styles.notch}}>
+          <path d='M0,1l1,-1l1,1Z' />
+        </svg>
+      </div>
+    </PopupPortal>;
+  }
+});
+
+export class TooltipTrigger extends React.Component {
+  props: any;
+  state: any;
+  id: string;
+
+  constructor(props) {
+    super(props);
+    this.state = { open: false };
+    this.id = `tooltip-trigger-${uniqueId('')}`;
+  }
+
+  render() {
+    const { children, content, ...props } = this.props;
+    const { open } = this.state;
+    if (!content) {
+      return children;
+    }
+    const child = React.Children.only(children);
+    return <React.Fragment>
+      {React.cloneElement(child, {
+        id: this.id,
+        onMouseEnter: (...args) => {
+          if (child.props.onMouseEnter) {
+            child.props.onMouseEnter(...args);
+          }
+          this.setState({open: true});
+        },
+        onMouseLeave: (...args) => {
+          if (child.props.onMouseLeave) {
+            child.props.onMouseLeave(...args);
+          }
+          this.setState({ open: false });
+        }
+      })}
+      {open && <Tooltip target={this.id} {...props}>{content}</Tooltip>}
+    </React.Fragment>;
+  }
+}
+
+export const Popup = flow(
+  onClickOutside,
+  withDynamicPosition()
+)(class PopupComponent extends React.Component {
+  static defaultProps = {
+    side: 'right'
+  };
+
+  props: any;
+
+  render() {
+    const {
+      children,
+      side,
+      elementRef,
+      dimensions: {
+        target,
+        element,
+        viewport
+      },
+      onClick
+    } = this.props;
+    const { position } = computePopupPosition({ side, target, element, viewport, gap: 10 });
+    return <PopupPortal>
+      <div
+        onClick={onClick}
+        ref={elementRef}
+        style={{
+          transform: `translate(${position.left}px, ${position.top}px)`,
+          ...styles.popup} as React.CSSProperties}
```

Comment:
Note: Having to do this cast everywhere is pretty rough. We should think about how we can avoid making this a standard pattern in the codebase when decide on a convention for our CSS.

Reply:
So... I just removed the cast and it worked fine and linted fine. Not sure why it was complaining before, but looks like it might not even be needed?

Reply:
Nevermind, strict test compilation fails without it...